File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        curSet = []
        def dfs(i, total):
            if total == target:
                res.append(curSet.copy())
                return 
            if i >= len(nums) or total > target:
                return
            
            curSet.append(nums[i])
            dfs(i, total + nums[i])

            curSet.pop()

            dfs(i + 1, total)

        dfs(0, 0)
        return res

File: Data Structures & Algorithms/subsets/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums) {
        let res = [];
        let subset = [];
        this.dfs(nums, 0, res, subset);
        return res;
    }
    dfs(nums, i, res, subset){
            if (i >= nums.length){
                res.push([...subset]);
                return;
            }
            subset.push(nums[i]);
            this.dfs(nums, i + 1, res, subset);
            subset.pop();
            this.dfs(nums, i + 1, res, subset);
     }
}
